### ISSUE TITLE
Investigate potential website performance improvements

### ISSUE THREAD

**Issue by c1505** [AgileVentures/WebsiteOne#1503]
Load on Events slows the entire site down.  There may be opportunity to improve performance.  I think a good first step is to understand what the hot spots are.  I don't have access to the performance information.  Any help is appreciated.

**Comment by arreche** [comment]
Line `<%= javascript_include_tag 'application', 'data-turbolinks-track' => true %>` in `app/views/layouts/_head.html.erb` file is adding an overhead to every page.

<img width="1280" alt="screen shot 2017-01-18 at 12 04 23" src="https://cloud.githubusercontent.com/assets/459813/22065354/cefc27ea-dd7f-11e6-9f78-a0cfccb8219f.png">

**Comment by tansaku** [comment]
we've adjusted the events page so it's not so big a load - I mean made the load time reasonable, by reducing the number of queries.

The problem before was when we had the individual event pages pinging the server with ajax requests to update the status of events from "off air" to "live", it killed the server.

This is not so much a problem with the existing site per se, but a problem we would encounter if we wanted all the event pages on users browser to be staying up to date in realtime.  We could try repeat pings with ajax again, but it seems like this is what websockets are for, which is why we're experimenting with websockets in the EventManager.

What we learn there might eventually come back to the monolith, or it might become a separate service ...

**Comment by arreche** [comment]
`javascript_include_tag` does't affect to the performance when the application is running in production mode.

By the way, MiniProfiler is an awesome profiling tool!
- http://railscasts.com/episodes/368-miniprofiler?autoplay=true

<img width="652" alt="screen shot 2017-02-03 at 23 35 33" src="https://cloud.githubusercontent.com/assets/459813/22612598/9bf71566-ea69-11e6-8d79-bb505481a493.png">

**Comment by tansaku** [comment]
we're experiencing the same issue again as mentioned by @mattlindsey in #1745 

we're seeing long load times in the event controller index action and loading the template:

![](https://www.dropbox.com/s/pue0dkn4373lcbb/Screenshot%202017-07-20%2009.37.31.png?dl=1)

**Comment by tansaku** [comment]
I've cloned production locally and am seeing this in the log:

```
Started GET "/events" for ::1 at 2017-07-20 09:14:21 +0000
Processing by EventsController#index as HTML
  Event Load (1.6ms)  SELECT "events".* FROM "events" WHERE "events"."category" = $1  [["category", "Scrum"]]
  Event Load (10.8ms)  SELECT "events".* FROM "events"
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=1003ms state=active
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=2005ms state=active
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=3003ms state=active
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=4004ms state=active
  Project Load (2.4ms)  SELECT "projects".* FROM "projects"
  EventInstance Load (4.0ms)  SELECT  "event_instances".* FROM "event_instances" WHERE "event_instances"."event_id" = $1 AND (updated_at BETWEEN '2017-07-19 09:14:40.991230' AND '2017-07-20 23:59:59.000000')  ORDER BY "event_instances"."updated_at" DESC LIMIT 1  [["event_id", 2]]
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=5005ms state=active
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=6005ms state=active
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=7004ms state=active
  EventInstance Load (4.5ms)  SELECT  "event_instances".* FROM "event_instances" WHERE "event_instances"."event_id" = $1 AND (updated_at BETWEEN '2017-07-19 09:14:59.017906' AND '2017-07-20 23:59:59.000000')  ORDER BY "event_instances"."updated_at" DESC LIMIT 1  [["event_id", 2640]]
  EventInstance Load (3.4ms)  SELECT  "event_instances".* FROM "event_instances" WHERE "event_instances"."event_id" = $1 AND (updated_at BETWEEN '2017-07-19 09:14:44.025157' AND '2017-07-20 23:59:59.000000')  ORDER BY "event_instances"."updated_at" DESC LIMIT 1  [["event_id", 3]]
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=8001ms state=active
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=9005ms state=active
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=10005ms state=active
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=11003ms state=active
source=rack-timeout id=90161fd1709063352ae00d66bd9d6940 timeout=200000000000ms service=12003ms state=active
  EventInstance Load (3.8ms)  SELECT  "event_instances".* FROM "event_instances" WHERE "event_instances"."event_id" = $1 AND (updated_at BETWEEN '2017-07-19 09:15:04.567178' AND '2017-07-20 23:59:59.000000')  ORDER BY "event_instances"."updated_at" DESC LIMIT 1  [["event_id", 2695]]
  EventInstance Load (4.4ms)  SELECT  "event_instances".* FROM "event_instances" WHERE "event_instances"."event_id" = $1 AND (updated_at BETWEEN '2017-07-19 09:15:04.574522' AND '2017-07-20 23:59:59.000000')  ORDER BY "event_instances"."updated_at" DESC LIMIT 1  [["event_id", 2544]]
  EventInstance Load (3.2ms)  SELECT  "event_instances".* FROM "event_instances" WHERE "event_instances"."event_id" = $1 AND (updated_at BETWEEN '2017-07-19 09:15:04.582654' AND '2017-07-20 23:59:59.000000')  ORDER BY "event_instances"."updated_at" DESC LIMIT 1  [["event_id", 2919]]
  EventInstance Load (3.2ms)  SELECT  "event_instances".* FROM "event_instances" WHERE "event_instances"."event_id" = $1 AND (updated_at BETWEEN '2017-07-19 09:26:34.589266' AND '2017-07-20 23:59:59.000000')  ORDER BY "event_instances"."updated_at" DESC LIMIT 1  [["event_id", 2791]]
  Rendered events/index.html.erb within layouts/application (8670.0ms)
  Rendered layouts/_meta_tags.html.erb (1.8ms)
  Rendered layouts/_head.html.erb (38.3ms)
  StaticPage Load (0.4ms)  SELECT  "static_pages".* FROM "static_pages" WHERE "static_pages"."title" = $1 LIMIT 1  [["title", "About Us"]]
  StaticPage Load (0.2ms)  SELECT  "static_pages".* FROM "static_pages" WHERE "static_pages"."title" = $1 LIMIT 1  [["title", "Pricing"]]
  StaticPage Load (0.3ms)  SELECT  "static_pages".* FROM "static_pages" WHERE "static_pages"."slug" = $1 LIMIT 1  [["slug", "Pricing"]]
  StaticPage Load (0.4ms)  SELECT  "static_pages".* FROM "static_pages" WHERE "static_pages"."title" = $1 LIMIT 1  [["title", "Getting started"]]
  Rendered layouts/_flash.html.erb (0.0ms)
  Rendered layouts/_navbar.html.erb (10.8ms)
  StaticPage Load (0.3ms)  SELECT  "static_pages".* FROM "static_pages" WHERE "static_pages"."title" = $1 LIMIT 1  [["title", "Sponsors"]]
  Rendered layouts/_sponsors.html.erb (7.5ms)
  CACHE (0.0ms)  SELECT  "static_pages".* FROM "static_pages" WHERE "static_pages"."title" = $1 LIMIT 1  [["title", "About Us"]]
  StaticPage Load (0.2ms)  SELECT  "static_pages".* FROM "static_pages" WHERE "static_pages"."title" = $1 LIMIT 1  [["title", "Getting Started"]]
  StaticPage Load (0.3ms)  SELECT  "static_pages".* FROM "static_pages" WHERE "static_pages"."slug" = $1 LIMIT 1  [["slug", "Getting Started"]]
  CACHE (0.0ms)  SELECT  "static_pages".* FROM "static_pages" WHERE "static_pages"."title" = $1 LIMIT 1  [["title", "Sponsors"]]
  Rendered layouts/_footer.html.erb (3.0ms)
Completed 200 OK in 12788ms (Views: 8718.4ms | ActiveRecord: 44.9ms)
```

**Comment by tansaku** [comment]
the number of database queries is reasonable (I think) that was the fix we made last time around ... I guess we've got some sort of rendering issue here to do with increased numbers of events/event_instances from the database, or something else that I can't guess ...

**Comment by tansaku** [comment]
@lokeshh @mattlindsey I'm just wondering if some recent change impacted this - will do a bisect locally ...

**Comment by tansaku** [comment]
it was actually much worse in the past ...

```
0198601a ---- Completed 200 OK in 31092ms (Views: 1687.9ms | ActiveRecord: 89.6ms)
```
^^^ this was before I fixed a boolean that was loading to many event instances

Since then it has been constant or slowly creeping up vvv
```
10bea298 ---- Completed 200 OK in 7075ms (Views: 3213.6ms | ActiveRecord: 42.9ms)
7704ce2a ----- Completed 200 OK in 7877ms (Views: 3418.9ms | ActiveRecord: 61.0ms)
master ---- Completed 200 OK in 7776ms (Views: 3581.9ms | ActiveRecord: 57.9ms)
```

**Comment by tansaku** [comment]
mmm - suggests we've got to look closer at the events controller and templates if we want to improve performance here ...

**Comment by tansaku** [comment]
at least the projects page loads nice and fast now we cleared out all the inactive projects  ...

**Comment by tansaku** [comment]
think I might have addressed this with #1778 and some other changes I made a while back to adjust the way in which the events page checks for live events
